lxd/identity/cache.go

```diff
@@ -0,0 +1,151 @@
+package identity
+
+import (
+	"crypto/x509"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/canonical/lxd/shared"
+	"github.com/canonical/lxd/shared/api"
+)
+
+// Cache represents an thread-safe in-memory cache of the identities in the database.
+type Cache struct {
+	entries map[string]map[string]CacheEntry
+	mu      sync.RWMutex
+}
+
+// CacheEntry represents an identity. The Certificate field is pre-computed for identities with AuthenticationMethod
+// api.AuthenticationMethodTLS.
+type CacheEntry struct {
+	Identifier           string
+	AuthenticationMethod string
+	IdentityType         string
+	Projects             []string
+	Certificate          *x509.Certificate
```

Comment:
Do we need to store a pointer of the cert here rather than just the cert struct?

Reply:
It isn't (or won't be) populated for all identity types so I figured a pointer will make it clear that it's optional. Happy either way.

Reply:
Ah yeah makes sense now.